app/src/main/java/com/futurice/freesound/feature/search/SoundItemViewModel.kt

```diff
@@ -0,0 +1,98 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.futurice.freesound.feature.search
+
+import com.futurice.freesound.feature.audio.AudioPlayer
+import com.futurice.freesound.feature.audio.PlaybackSource
+import com.futurice.freesound.feature.audio.PlayerState
+import com.futurice.freesound.feature.audio.from
+import com.futurice.freesound.feature.common.Navigator
+import com.futurice.freesound.network.api.FreeSoundApiService
+import com.futurice.freesound.network.api.model.Sound
+import com.futurice.freesound.viewmodel.SimpleViewModel
+import com.google.auto.factory.AutoFactory
+import com.google.auto.factory.Provided
+import io.reactivex.Observable
+import io.reactivex.Single
+import polanski.option.Option
+import java.text.DateFormat
+
+@AutoFactory
+internal class SoundItemViewModel(private val sound: Sound,
+                                  private @Provided val navigator: Navigator,
+                                  private @Provided val audioPlayer: AudioPlayer,
+                                  private @Provided val freeSoundApiService: FreeSoundApiService) : SimpleViewModel() {
+
+    private val thumbnail: String
+        get() = sound.images.medSizeWaveformUrl
+
+    private val currentPercentage: Observable<Option<Int>>
+        get() = audioPlayer.timePositionMsOnceAndStream
+                .map { positionMs -> toPercentage(positionMs, sound.duration) }
+                .map { Option.ofObj(it) }
+
+    fun thumbnailImageUrl(): Single<String> = Single.just(thumbnail)
+
+    fun name(): Single<String> = Single.just(sound.name)
+
+    fun userAvatar(): Single<String> =
+            freeSoundApiService.getUser(sound.username)
+                    .map { (_, _, avatar) -> avatar.medium }
+                    .cache()
+
+    fun createdDate(): Single<String> =
+            Single.just(sound.created)
+                    .map { it.time }
+                    .map { d -> DateFormat.getDateInstance().format(d) }
+
+    fun username(): Single<String> = Single.just(sound.username)
+
+    fun description(): Single<String> = Single.just(sound.description)
+
+    fun duration(): Single<Int> =
+            Single.just(sound.duration)
+                    .map { duration -> Math.ceil(duration.toDouble()).toInt() }
+                    .map { duration -> Math.max(duration, 1) }
+
+    fun progressPercentage(): Observable<Option<Int>> =
+            audioPlayer.playerStateOnceAndStream
+                    .switchMap(this::progressOrNothing)
+
+    fun openDetails() {
+        navigator.openSoundDetails(sound)
+    }
+
+    fun toggleSoundPlayback() {
+        audioPlayer.togglePlayback(
+                PlaybackSource(from(sound.id),
+                        sound.previews.lowQualityMp3Url))
+    }
+
+    private fun progressOrNothing(playerState: PlayerState): Observable<Option<Int>> {
```

Comment:
Thoughts on making these helper functions be extension methods instead?

Reply:
Could you give an example how it would make it better?

Reply:
Only that a private extension method might read better from the call site, than when the receiver is a parameter:

So they would look a bit like:
`playerState.progressOrNothing()` 
`playerState.source.isThisSound()`
`position.ToPercentage()`

Reply:
For private extensions there's not much of a downside in terms of perceived instance method pollution.